x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/EsqlArithmeticOperation.java

```diff
@@ -168,4 +169,10 @@ public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEv
         }
         throw new EsqlIllegalArgumentException("Unsupported type " + leftType);
     }
+
+    @Override
+    public Nullability nullable() {
+        return Nullability.TRUE;
```

Comment:
Is this right? For something like `| eval x = 1, y = 2 | eval z = x + y | where z is not null`, the nullability of the addition should still be `FALSE`, right?

Reply:
I don't think so. You'll know that the result is not null only after you fold it, but nullability does not rely on folding.

A practical example is `1.7976931348623157E+308 + 1.7976931348623157E+308`: both operands are not null, but the result is null (infinity)

Reply:
> You'll know that the result is not null only after you fold it, but nullability does not rely on folding.

That of `Literal`s implicitly does, right? `Literal#nullable()` literally checks for `null`.
However, more generally, my intuition is that for expressions with literals only, when we can fold them, we should be able to establish the exact nullability, rather than be conservative (i.e. return the "worst": `Nullability.TRUE`). And this could manifest in the inter-play of the optimisation rules, not necessarily extend other `nullable()`s to check for values.

Otherwise, replacing `nullable()`s with `TRUE`s will make `FoldNull` effectively simply check if there's a `null` literal in the tree.

Also, when we'll get to broader MVs support, comparisons and other boolean predicates (comparisons, `LIKE`, `startsWith()` etc.)  can keep their children-delegating logic.

(However, your proposed solution is probably still a better intermediary step than current state.)

Reply:
> we should be able to establish the exact nullability

I don't think it's about nullability. If an expression is foldable, we have optimization rules that will do it for us and everything will work just fine, we just have to avoid messing up with nullability.

And consider that all this applies also to non-foldable expressions:

```
FROM idx | WHERE coalesce(a, 10) + coalesce(b, 20) IS NOT NULL
```

in this case you cannot even fold, because a and b are not foldable, but you can easily say that `coalesce(a, 10)` and `coalesce(a, 10)` are Nullable.FALSE.

I think reasoning on folding is a red herring here.

Reply:
> I don't think nullability and folding to null should be mixed.

Isn't that what `FoldNull` does: it folds to `null` based on nullability, tho just for `IsNotNull`/`IsNull`; and effectively, the proposed fix deactivates this check.

> I think reasoning on folding is a red herring here.

And the fix this PR proposes should extend to `coalesce()` too, besides the `EsqlArithmeticOperation`s, I think you're right.

However, the issue with comparisons &co. above stands. If the comparison would be folded before `FoldNull` gets to it, the change in this PR wouldn't be required anymore.

Besides the MV issue: we don't do this now, but if we'd have `FieldAttribute`s' nullability set based on type and we'd compare two `constant_keyword`s, the `isNull()` _should_ be optimised away. With the proposed fix, this wouldn't happen anymore.